cargo-registry-index/lib.rs

```diff
@@ -496,9 +517,37 @@ impl Repository {
 
     fn fetch_options(credentials: &Credentials) -> git2::FetchOptions<'_> {
         let mut callbacks = git2::RemoteCallbacks::new();
+
         callbacks.credentials(move |_, user_from_url, cred_type| {
             credentials.git2_callback(user_from_url, cred_type)
         });
+
+        callbacks.certificate_check(|cert, hostname| {
+            if hostname != "github.com" {
+                return Ok(CertificateCheckStatus::CertificatePassthrough);
+            }
+
+            let Some(hostkey) = cert.as_hostkey() else {
+                // if this is a X.509 certificate then we rely on the default
+                // certificate checks in libgit2
+                return Ok(CertificateCheckStatus::CertificatePassthrough)
+            };
+
+            let Some(hash) = hostkey.hash_sha256() else {
+                let mut error = git2::Error::from_str("Failed to read SHA256 hash of the SSH hostkey certificate");
+                error.set_code(ErrorCode::Certificate);
+                return Err(error)
+            };
+
+            if GITHUB_FINGERPRINTS.iter().any(|fingerprint| hash == fingerprint.as_bytes()) {
+                Ok(CertificateCheckStatus::CertificateOk)
+            } else {
+                let mut error = git2::Error::from_str("Certificate check for github.com failed");
+                error.set_code(ErrorCode::Certificate);
+                Err(error)
+            }
```

Comment:
This seems... fine, even though I'd do something like:

```rust
if hostname == "github.com" {
    if let Some(sha256) = cert.as_hostkey().and_then(|key| key.hash_sha256()) {
        if GITHUB_FINGERPRINTS.iter().any(|fingerprint| hash == fingerprint.as_bytes()) {
            return Ok(CertificateCheckStatus::CertificateOk);
        }
    }
}
Ok(CertificateCheckStatus::CertificatePassthrough)
```

That way we don't override the libgit2 behavior if GitHub ever rotates keys.

Reply:
✅